src/lib.rs

```diff
@@ -59,6 +66,49 @@ pub struct Pointer<T> {
     pending: PendingRef,
 }
 
+impl<T> Pointer<T> {
+    /// Creates a new `WeakPointer` to this component.
+    pub fn downgrade(&self) -> WeakPointer<T> {
+        let epoch = self.target.read().unwrap().epoch[self.index];
+        WeakPointer {
+            index: self.index,
+            epoch: epoch,
+            target: Arc::downgrade(&self.target),
+            pending: Arc::downgrade(&self.pending),
+        }
+    }
+}
+
+/// Weak variant of `Pointer`.
+pub struct WeakPointer<T> {
+    index: usize,
+    epoch: Epoch,
+    target: WeakStorageRef<T>,
+    pending: WeakPendingRef,
+}
+
+impl<T> WeakPointer<T> {
+    /// Upgrades the `WeakPointer` to an `Pointer`, if possible.
+    /// Returns `None` if the strong count has reached zero and the inner value was destroyed.
+    pub fn upgrade(&self) -> Option<Pointer<T>> {
+        match self.target.upgrade().and(self.pending.upgrade()) {
+            None => None,
+            Some(arc) => {
+                let target = self.target.upgrade().unwrap();
+                if target.read().unwrap().meta[self.index] == 0 ||
```

Comment:
Well, interesting proposal.
I think `upgrade` as `Storage` method makes no sense - you still need lock.
As the most often use case will be accessing `Pointer` soon after `upgrade`ing it, we can add new methods for `Lock`s: `access_weak` and `access_weak_mut` which will return `Result<Pointer<T>>`. But we still need regular `upgrade` method.